trestle/core/control_reader.py

```diff
@@ -0,0 +1,849 @@
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Handle reading of writing controls from markdown."""
+import logging
+import pathlib
+import string
+from datetime import datetime
+from typing import Any, Dict, List, Optional, Tuple
+
+import frontmatter
+
+import trestle.core.generic_oscal as generic
+import trestle.oscal.catalog as cat
+import trestle.oscal.common as com
+from trestle.common import const
+from trestle.common.err import TrestleError
+from trestle.common.list_utils import as_list, delete_list_from_list, none_if_empty
+from trestle.common.model_utils import ModelUtils
+from trestle.common.str_utils import spaces_and_caps_to_snake
+from trestle.core import generators as gens
+from trestle.core.control_interface import CompDict, ComponentImpInfo, ContextPurpose, ControlContext, ControlInterface
+from trestle.core.markdown.markdown_api import MarkdownAPI
+from trestle.core.markdown.markdown_processor import MarkdownNode
+from trestle.oscal import common
+from trestle.oscal import component as comp
+from trestle.oscal import profile as prof
+
+logger = logging.getLogger(__name__)
+
+
+class ControlReader():
+    """Class to read controls from markdown."""
+
+    @staticmethod
+    def _load_control_lines_and_header(control_file: pathlib.Path) -> Tuple[List[str], Dict[str, Any]]:
+        lines: List[str] = []
+        try:
+            content = control_file.open('r', encoding=const.FILE_ENCODING).read()
+        except UnicodeDecodeError as e:
+            logger.debug(f'See: {const.WEBSITE_ROOT}/errors/#utf-8-encoding-only')
+            raise TrestleError(f'Unable to load file due to utf-8 encoding issues: {e}')
+        try:
+            fm = frontmatter.loads(content)
+        except Exception as e:
+            logger.error(
+                f'Error parsing yaml header from file {control_file}. '
+                f'This is most likely due to an incorrect yaml structure.'
+            )
+            raise TrestleError(f'Failure parsing yaml header on file {control_file}: {e}')
+        raw_lines = fm.content.split('\n')
+        header = fm.metadata
+        # Any fully blank lines will be retained but as empty strings
+        lines = [line.strip('\r\n').rstrip() for line in raw_lines]
+        clean_lines = []
+        # need to keep indentation and empty lines
+        for line in lines:
+            if line.startswith('<!--') or line.startswith('__________________'):
+                continue
+            clean_lines.append(line)
+        return clean_lines, header
```

Comment:
This part ideally should be done via `MarkdownNode`, but definitely too much for this PR. So note to self.

Reply:
yes - hope this can be reworked.  no changes made.